document/AnnexE-Determinism

```diff
@@ -0,0 +1,136 @@
+
+For a tool to produce deterministic output, it must emit array and dictionary elements
+in a deterministic order.
+
+For some arrays, the SARIF format requires a specific ordering.
+For example, within the `stack.Frames` property, SARIF requires the `annotatedCodeLocation` object
+representing the most deeply nested function call to appear first.
+
+For other arrays, the SARIF format does not require a specific ordering.
+For example, within the `file.hashes` property, SARIF does not require the `hash` objects
+to appear in any particular order.
+For such arrays, a tool can ensure the order by sorting the array elements before
+writing them to the log file.
+For example, it might sort the `hash` objects alphabetically by the string value of the
+`hash.algorithm` property.
+
+A tool might similarly choose to emit the string elements of a `properties.tags` array in
+locale-insensitive alphabetical order.
+
+The array of `result` objects presents more of a problem.
+A multi-threaded analysis tool analyzing multiple files in parallel might produce results
+in any order, and there is no natural order for the results.
+A tool might choose to order them, for example, first alphabetically by analysis target URI,
+then numerically by line number, then by column number, then alphabetically by rule id.
+
+For dictionaries such as the `run.rules` object or the `run.files` object,
+a tool might order the property names alphabetically, using a locale-insensitive ordering.
+
+## Case-insensitive file format elements { #Annex-E-Case }
+
+Case-insensitive file systems can also cause non-deterministic output.
+Suppose a URI-valued property contains a file path from a file system that is
+case-preserving but case-insensitive (such as NTFS):
+
+```
+"uri": "file:///C:/Code/File1.c"
+```
+
+Suppose the file is then renamed to `file.c`. NTFS treats `File.c` and `file.c` as the same file.
+But a subsequent run of the tool might produce
+
+```
+"uri": "file:///C:/Code/file1.c"
+```
+
+A tool might choose to ignore this issue, since this would happen rarely, and the only
+consequence of ignoring it is that a subsequent step in a build process might execute
+even though it does not strictly have to.
+
+To produce absolutely deterministic output, the tool might instead choose to pull all
+file paths to lowercase or to uppercase, using a locale-insensitive case conversion:
+
+```
+"uri": "file:///C:/code/file1.c"
+```
+
+~IsoNote
+
+On the .NET platform, the recommendation is to pull strings to uppercase
+using the method `String.ToUpperInvariant`, which would result in:
+
+```
+"uri": "file:///C:/CODE/FILE1.C"
+```
+~
+
+## Post-processing { #Annex-E-PostProcessing }
+
+Even if an analysis tool does not produce deterministic output, a post-processing tool could
+produce deterministic output by removing or ignoring non-deterministic elements,
+reordering array elements and object properties,
+and adjusting the casing of case-insensitive string values
```

Comment:
more typically, you'd perform case-insensitive comparisons, rather than adjust the casing in a post-processing step. also, you need a period on the end of this sentence.